### PR TITLE
stacked on v2-ack: add upd req+resp with workqueue

### PR DESCRIPTION
**Pull request by jw-msft**
- Add upd mgmt and request operation state machine
- Publish upd_req message to agent topic
- Wait for and verifies PUBACK from mqtt broker
- Wait for upd_resp message in OnMessage callbacks
- Create update and reporting work queues
- Push the response json message body onto update work queue
- Clang format and cmake format
- Add enum to str functions

### CODE REVIEW

File: src/extensions/agent_modules/shared/common/inc/aduc/adu_upd.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * @brief THe update state of the the sending of 'upd_req' and handling of the 'upd_resp' response as per the adu protocol.
+ */
+typedef enum tagADU_UPD_STATE
+{
+    ADU_UPD_STATE_READY             =  0, /**< The update module is ready to start requesting in the next frame of execution. */
+    ADU_UPD_STATE_IDLEWAIT          =  1, /**< The upd state is waiting for wait timer timeout to poll for updates. */
+    ADU_UPD_STATE_REQUESTING        =  2, /**< In the process of requesting for updates. */
+    ADU_UPD_STATE_RETRYWAIT         =  3, /**< The upd state is waiting for wait timer timeout to poll for updates. */
+    ADU_UPD_STATE_REQUEST_ACK       =  4, /**< Received PUBACK from the MQTT broker and is awaiting a response message. */
+    ADU_UPD_STATE_PROCESSING_UPDATE =  6, /**< In the process of processing the update for instation on the device. */
```
Comment (Nox-MSFT) [inline src/extensions/agent_modules/shared/common/inc/aduc/adu_upd.h:19]: typo: instation => installation

File: src/extensions/agent_modules/shared/upd_utils/src/adu_upd_utils.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+/**
+ * @brief Gets the update workqueue handle from the callback user object.
+ * @param obj The callback's user object.
+ * @return ADUC_Retriable_Operation_Context* The retriable operation context.
+ */
+WorkQueueHandle WorkQueueHandleFromCallbackUserObj(void* obj)
+{
+    ADUC_MQTT_CLIENT_MODULE_STATE* ownerModuleState = (ADUC_MQTT_CLIENT_MODULE_STATE*)obj;
+    ADUC_WorkQueues* workQueues = (ADUC_WorkQueues*)(ownerModuleState->moduleInitData);
+    return workQueues->updateWorkQueue;
```
Comment (Nox-MSFT) [inline src/extensions/agent_modules/shared/upd_utils/src/adu_upd_utils.c:39]: Suggestion: NULL check and gracefully return null? 

Same in 2 functions below.

**Review by Nox-MSFT — APPROVED** [review]
Approved with suggestions.

File: src/utils/workqueue/inc/aduc/workqueue_item.h [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+#include <aduc/c_utils.h>
+#include <time.h>
+
+typedef void* WorkQueueItemHandle;
+
+EXTERN_C_BEGIN
+
+time_t WorkQueueItem_GetTimeAdded(WorkQueueItemHandle handle);
+char* WorkQueueItem_GetUpdateResultMessageJson(WorkQueueItemHandle handle);
```
[the remark this answers is not in the document]
Comment (jw-msft) [reply src/utils/workqueue/inc/aduc/workqueue_item.h]: Fixed with commit: [ren to more generic _GetData](https://github.com/Azure/iot-hub-device-update/pull/585/commits/6e205a3e49c40a604051d6da1e72e73a03de10f8)

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_update_management.c [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+void OnMessage_upd_resp(
+    struct mosquitto* mosq, void* obj, const struct mosquitto_message* msg, const mosquitto_property* props)
+{
+    char* correlationData = NULL;
+    uint16_t correlationDataByteLen = 0;
+
+    WorkQueueHandle updateWorkQueueHandle = WorkQueueHandleFromCallbackUserObj(obj);
+    ADUC_Retriable_Operation_Context* retriableOperationContext = RetriableOperationContextFromCallbackUserObj(obj);
+    ADUC_Update_Request_Operation_Data* updateData =
+        retriableOperationContext == NULL ? NULL : UpdateDataFromRetriableOperationContext(retriableOperationContext);
```
[the remark this answers is not in the document]
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_update_management.c]: fixed with commit: [rmv ternary NULL chk for graceful handling from PR commit lower in stack](https://github.com/Azure/iot-hub-device-update/pull/585/commits/aab66a0ea5514acea0a95a19e735785affb7d95a)

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_update_management.c [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+    if (updateWorkQueueHandle == NULL || retriableOperationContext == NULL || updateData == NULL)
+    {
+        Log_Error("invalid callback user obj");
+        goto done;
+    }
+
+    json_print_properties(props);
```
[the remark this answers is not in the document]
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_update_management.c:87]: opened tracking bug: [Bug 48195424](https://microsoft.visualstudio.com/OS/_workitems/edit/48195424): [Gen2] Make loglevel part of du-config.json

File: src/extensions/agent_modules/shared/common/inc/aduc/adu_upd.h (hunk 1 above)
Comment (jw-msft) [reply src/extensions/agent_modules/shared/common/inc/aduc/adu_upd.h]: Fixed with commit: [fix typo - instation -> installation](https://github.com/Azure/iot-hub-device-update/pull/585/commits/42862b83006616fb69b711480a61e69dc24ab099)

File: src/extensions/agent_modules/shared/upd_utils/src/adu_upd_utils.c (hunk 2 above)
Comment (jw-msft) [reply src/extensions/agent_modules/shared/upd_utils/src/adu_upd_utils.c]: fixed with commit: [handle nulls](https://github.com/Azure/iot-hub-device-update/pull/585/commits/aa0f584e401c72e2c5ed5e37ebbec479659e6ae0)

**Comment by jw-msft** [comment]
merged up to top pr stack that was merged to topic

**Pull request closed without merging**